### PR TITLE
TIDO v5: show panels wrapper which contains text and switch between different types of text content

### PR DESCRIPTION
**Pull request by orlinmalkja**

### CODE REVIEW

**Comment by orlinmalkja** [comment]
@paulpestov Shall we display by default the first text content type or should one make the initial display configurable ? 
I think it is not necessary to make it configurable, but just show initially the first text content type, it seems natural to me. The reason against making it configurable, is that obtaining the desired content type can be achieved by a very simple action - clicking at another button.

**Comment by paulpestov** [comment]
Somehow I get this error at `npm run dev`:

  primeicons/primeicons.css (imported by .../tido/src/App.tsx)
  primereact/button (imported by .../tido/src/components/panel/TextTypes.tsx)

File: src/main.tsx [hunk 1]
```diff
@@ -1,9 +1,9 @@
-import { StrictMode } from 'react'
-import { createRoot } from 'react-dom/client'
-import App from './App.tsx'
+import { StrictMode } from "react";
```
Comment (paulpestov) [inline src/main.tsx:1]: I suggest to keep the whole app without semicolons and with single quotes instead of double quotes. It just looks cleaner in my opinion. For this there are ESLint rules that we can add in the config. If you then apply the "lint on save" feature in your IDE it would automatically pick up the new rules.

File: tsconfig.json [hunk 2]
```diff
@@ -15,12 +14,17 @@
...
+
     /* Linting */
     "strict": true,
     "noUnusedLocals": true,
     "noUnusedParameters": true,
     "noFallthroughCasesInSwitch": true,
     "noUncheckedSideEffectImports": true
   },
-  "include": ["src"]
+  "include": ["./"]
```
Comment (paulpestov) [inline tsconfig.json:29]: Why this change?

File: src/components/panel/Panel.tsx [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+// prop: url - should be the url of collection or manifest
+const Panel: FC = ({ url }) => {
+  const { config, setConfig } = useContext(ConfigContext);
+  const [data, setData] = useState();
+  const [itemUrl, setItemUrl] = useState("");
+  const [text, setText] = useState<React.ReactNode | undefined>();
+  const [textTypes, setTextTypes] = useState([]);
+  const [activeText, setActiveText] = useState("");
+
+  async function getItemUrl(documentData): string {
```
Comment (paulpestov) [inline src/components/panel/Panel.tsx:20]: Type?

File: src/components/panel/Panel.tsx [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+    if (itemData["content"].length === 0) return;
+
+    const content = itemData["content"];
+    for (let i = 0; i < content.length; i++) {
+      types.push(getContentType(content[i].type));
+    }
+    setTextTypes(() => types);
+  }
+
+  function getContentType(value): string {
```
Comment (paulpestov) [inline src/components/panel/Panel.tsx:46]: Type?

**Comment by orlinmalkja** [comment]
> Somehow I get this error at `npm run dev`:
> 
> Error: The following dependencies are imported but could not be resolved: primeicons/primeicons.css (imported by .../tido/src/App.tsx) primereact/button (imported by .../tido/src/components/panel/TextTypes.tsx)

have you installed primereact locally ?

**Comment by paulpestov** [comment]
I think its missing in package.json

File: tsconfig.json (hunk 2 above)
Comment (orlinmalkja) [reply tsconfig.json]: restored it back.

File: src/components/panel/Panel.tsx (hunk 3 above)
Comment (orlinmalkja) [reply src/components/panel/Panel.tsx]: added the type

File: src/components/panel/Panel.tsx (hunk 4 above)
Comment (orlinmalkja) [reply src/components/panel/Panel.tsx]: added the type

**Comment by paulpestov** [comment]
I had errors during the installation again, this time it was the SASS compiler. In any case, I think we don't need SASS anymore since it was mainly used for some global styling that we can solve otherwise. Also the nesting approach is becoming a native CSSS standard (https://caniuse.com/css-nesting). I'm already making use of it in the `preflight.css`.

File: src/App.tsx [hunk 5]
```diff
@@ -1,9 +1,34 @@
-function App() {
+import PanelsWrapper from './components/PanelsWrapper';
+import { useEffect, useState } from 'react';
+import { useConfig} from '@/contexts/ConfigContext'
+
+
+
+function App({customConfig}) {
```
Comment (paulpestov) [inline src/App.tsx:7]: Type?

File: src/App.tsx [hunk 6]
```diff
@@ -1,9 +1,34 @@
...
+
+  const [loading, setLoading] = useState(true)
+  const { config, updateConfig} = useConfig()
```
Comment (paulpestov) [inline src/App.tsx:10]: updateConfig not exported, code style

File: src/contexts/ConfigContext.tsx [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+  );
+};
+
+function useConfig(): ConfigProvider {
+  return useContext(ConfigContext)
+}
+
+
+
+export { ConfigProvider, useConfig };
```
Comment (paulpestov) [inline src/contexts/ConfigContext.tsx:39]: The export gives me a fast fresh ESLint error. Since the official React [example](https://codesandbox.io/p/sandbox/react-dev-59fz7l?file=%2Fsrc%2FTasksContext.js&utm_medium=sandpack) shows to use it in one file I suggest to add supress the `react-refresh/only-export-components` ESlint rule.

File: src/contexts/ConfigContext.tsx [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+const ConfigContext = createContext({});
+
+interface ConfigProvider {
+  config?: Config,
+  setConfig?: () => void,
+  openedPanels?: Panel[],
+  setOpenedPanels?: (panels: Panel[]) => void
+}
+
+interface ConfigProviderProps {
```
Comment (paulpestov) [inline src/contexts/ConfigContext.tsx:12]: `..Props {` Here it shows an irregular whitespace as `NBSP` for me

File: src/App.tsx [hunk 9]
```diff
@@ -1,9 +1,34 @@
...
+  useEffect(() => {
+    // load Config
+    // TODO: url config will need to get read
+    function loadConfig() {
+      const totalConfig = {...config, ...customConfig}
+      updateConfig(totalConfig)
+      setLoading(false)
+    }
+    loadConfig()
+  }, [])
```
Comment (paulpestov) [inline src/App.tsx:21]: Shouldn't the dependencies array contain all variables it is using inside?

File: src/components/PanelsWrapper.tsx [hunk 10]
```diff
@@ -0,0 +1,36 @@
+import { FC, useEffect } from 'react';
+import Panel from '@/components/panel/Panel';
```
Comment (paulpestov) [inline src/components/PanelsWrapper.tsx:2]: I think this is imported accidentally. Instead it should come from types.d.ts

File: src/components/PanelsWrapper.tsx [hunk 11]
```diff
@@ -0,0 +1,36 @@
...
+import { useConfig } from '@/contexts/ConfigContext';
+import { getManifestUrl, getCollectionUrl } from '@/utils/panel';
+
+const PanelsWrapper: FC = ({}) => {
+  const { config, openedPanels, setOpenedPanels } = useConfig();
+
+  function initOpenedPanels(panels: Panel[]) {
```
Comment (paulpestov) [inline src/components/PanelsWrapper.tsx:9]: Is this wrapper function needed? You could just call `setOpenedPanels` directly.

File: src/types.d.ts [hunk 12]
```diff
@@ -178,6 +183,10 @@ declare global {
         info: string,
         warning: string
     }
+    interface Panel {
```
Comment (paulpestov) [inline src/types.d.ts:186]: We should rename this to `PanelConfig` because we have a component with the same name.

File: src/components/PanelsWrapper.tsx [hunk 13]
```diff
@@ -0,0 +1,36 @@
...
+    }
+    initOpenedPanels(config.panels);
+  }, []);
+
+  
+  const panels = openedPanels ?
+    openedPanels.length > 0 &&
+    openedPanels.map((panel: Panel, i: number) => (
+      <div key={i} className="t-mr-6">
+        <Panel url={ panel.collection ? getCollectionUrl(panel): getManifestUrl(panel)} />
```
Comment (paulpestov) [inline src/components/PanelsWrapper.tsx:28]: Just pass the whole `PanelConfig` as prop here and let the component figure out what it should load. This way we can also read other upcoming config options inside it.

File: src/components/panel/Panel.tsx [hunk 14]
```diff
@@ -0,0 +1,147 @@
...
+  const [error, setError] = useState<boolean | string>(false)
+  const [loading, setLoading] = useState<boolean>(true)
+
+
+  async function getItemUrl(documentData: Manifest | Collection): Promise<string | null> {
+    // if collection - then we should read the api data from the manifest and get its first sequence item id
+    // if manifest - we retrieve the first sequence item id
+
+    // panel in the config having a document with the prop url
+    const panel: Panel | undefined = getPanel(url, config)
```
Comment (paulpestov) [inline src/components/panel/Panel.tsx:32]: This should become obsolete as soon as you pass the PanelConfig as prop.

File: src/App.tsx (hunk 6 above)
Comment (orlinmalkja) [reply src/App.tsx]: I think updateConfig is not necessary, since setConfig itself in the ConfigContext can receive a new Config as an argument and update the config. Therefore, I removed updateConfig. This should still be fine, or ?

File: src/App.tsx (hunk 5 above)
Comment (orlinmalkja) [reply src/App.tsx]: converted the component to FC component and used a typescript interface for its type.

File: src/contexts/ConfigContext.tsx (hunk 7 above)
Comment (orlinmalkja) [reply src/contexts/ConfigContext.tsx:36]: In my code eslint does not raise any error.. Is it still the case for you ?

File: src/App.tsx (hunk 9 above)
Comment (orlinmalkja) [reply src/App.tsx:23]: I am not sure, when have a config change which affects only a specific component, then we would rerender the App component and its children, right ?  

Moreover, the loadConfig in App.tsx will be executed at the very first beginning.

File: src/components/PanelsWrapper.tsx (hunk 10 above)
Comment (orlinmalkja) [reply src/components/PanelsWrapper.tsx:2]: Actually it is meant to import the Panel component and not the PanelConfig (I updated the name in types.d.ts for the Panel config).

File: src/components/PanelsWrapper.tsx (hunk 11 above)
Comment (orlinmalkja) [reply src/components/PanelsWrapper.tsx]: yes, true. I removed initOpenedPanels()

File: src/types.d.ts (hunk 12 above)
Comment (orlinmalkja) [reply src/types.d.ts]: yes, sounds good, I renamed it.

File: src/utils/http.tsx [hunk 15]
```diff
@@ -0,0 +1,21 @@
+
+export async function get(url: string): Promise<Manifest | Collection | null> {
+  let response: Manifest | Collection | null = null
```
Comment (paulpestov) [inline src/utils/http.tsx:3]: This is not generic I'm afraid. Mentioning specific interfaces and using specific error messages should be avoided.

File: src/config.tsx [hunk 16]
```diff
@@ -0,0 +1,15 @@
+const defaultConfig: Config = {
+  globalTree: true,
+  panels: [
+    {
+      entrypoint:
```
Comment (paulpestov) [inline src/config.tsx:5]: Hm I thought you would implement this in some next PR. This was just an idea, I'm not sure about this yet.

File: index.html [hunk 17]
```diff
@@ -27,8 +27,19 @@
...
   window.addEventListener('load', function () {
-    window.tido = new window.Tido();
-  });
+    window.tido = new window.Tido({
+      globalTree: true,
+      panels: [
+        {
+          entrypoint:
+            'https://api.ahiqar.sub.uni-goettingen.de/textapi/ahiqar/syriac/collection.json',
+          m: 3,
```
Comment (paulpestov) [inline index.html:36]: Hm I thought you would implement this in some next PR. This was just an idea, I'm not sure about this yet.

File: src/App.tsx (hunk 9 above)
Comment (orlinmalkja) [reply src/App.tsx:26]: Actually I think that useEffect can make it possible to execute something only once. If I place code outside of useEffect it will be executed repeatedly. 

I removed useEffect() from PanelsWrapper, but to my opinion it seems like it is necessary in other components.

File: src/components/PanelsWrapper.tsx (hunk 13 above)
Comment (orlinmalkja) [reply src/components/PanelsWrapper.tsx]: Yes it is a good idea to pass the whole PanelConfig as a prop. I did it.

File: src/utils/http.tsx (hunk 15 above)
Comment (orlinmalkja) [reply src/utils/http.tsx]: Yes, in my latest commit I try to create a generic function to fetch api data. On that commit I try to parse the data based on the content type which I get from the headers of the response received. Regarding the content types of the response I had a look on this [article](https://medium.com/@anatoliiyatsenko/understanding-fetch-api-response-methods-and-the-content-type-header-6dcbe7b24ded)

File: src/App.tsx (hunk 9 above)
Comment (paulpestov) [reply src/App.tsx:26]: Hm I still see the useEffect in PanelsWrapper. Also why do you need useEffects in other components?

**Review by paulpestov — APPROVED** [review]

File: src/main.tsx [hunk 18]
```diff
@@ -5,8 +5,14 @@ import './css/style.css'
 
 import App from './App.tsx'
 
-window.Tido = function Tido(config = {}) {
-    createRoot(document.getElementById('app')!).render(<App customConfig={config} />)
+declare global {
```
Comment (orlinmalkja) [inline src/main.tsx:8]: @paulpestov I merged the PR, but I have still one question. Which is the idea behind the lines 8-12 in this file ?

Comment (paulpestov) [reply src/main.tsx:8]: Just to satisfy Typescript
